src/MessagePack.UnityClient/Assets/Scripts/MessagePack/MessagePackSerializer.NonGeneric.cs

```diff
@@ -220,38 +220,21 @@ internal CompiledMethods(Type type)
 
                 {
                     // public static void Serialize<T>(ref MessagePackWriter writer, T obj, MessagePackSerializerOptions options)
-                    MethodInfo serialize = GetMethod(nameof(Serialize), type, new Type[] { typeof(MessagePackWriter).MakeByRefType(), null, typeof(MessagePackSerializerOptions) });
+                    MethodInfo serialize = GetMethod(nameof(SerializeSemiGeneric), type, new Type[] { typeof(MessagePackWriter).MakeByRefType(), typeof(Object), typeof(MessagePackSerializerOptions) });
```

Comment:
Is there any performance penalty in SemiGeneric? Should it be only used under ENABLE_IL2CPP then, where is no other option?

Reply:
From my point of view it should not be any perfomance penalty.
If you check SerializeSemiGeneric and DeserializeSemiGeneric methods -- they are just wrappers around current implementation with assitional casting. Do not think that casting will add any significatn penalty.

But I cannot garantee it.
Do you have any benchmarks? Can you compare two implementations?

Reply:
It seems that the delegate creates takes an object as command line argument anyway, so there is no penalty there. But calling Serialize/Deserialize now introduces unnecessary boxing for value types. Is it executed anywhere in a hot path and critical?

Reply:
I changed a bit code.
Now **SerializeSemiGeneric** and **DeserializeSemiGeneric** are using only inside MessagePackSerializer.NonGeneric.cs file.
So it decreases cases where additional casting is using.

Do not know how methods from MessagePackSerializer.NonGeneric.cs are used in MessagePack (I am not MessagePack author).
So cannot answer on your question "Is it executed anywhere in a hot path and critical?"

Reply:
RE: ENABLE_IL2CPP 

I do not know what ENABLE_IL2CPP is for. 
When this derective is using? What is advantages/disadvantages to use this option?
And I guess MessagePack nuget package will be build without this option, will not it?

Reply:
The ENABLE_IL2CPP is for Unity3D game engine where ahead of time compilation renders Emit code generation unusable.

Reply:
I think making the change only on unity wouldn't fix the original issue, which was an app running on iOS.